greenmail-core/src/main/java/com/icegreen/greenmail/imap/commands/SortCommand.java

```diff
@@ -0,0 +1,326 @@
+            } catch (MessagingException e) {
+                throw new RuntimeException(e.getMessage(), e);
+            }
+        }
+
+        private String getFrom(StoredMessage message) throws MessagingException {
+            Address[] addresses = message.getMimeMessage().getFrom();
+            if (addresses == null || addresses.length == 0) {
+                return null;
+            }
+            return String.valueOf(addresses[0]);
+        }
+
+        private String getTo(StoredMessage message) throws MessagingException {
+            Address[] addresses = message.getMimeMessage().getRecipients(Message.RecipientType.TO);
+            if (addresses == null || addresses.length == 0) {
+                return null;
+            }
+            return String.valueOf(addresses[0]);
+        }
+
+        private String getCc(StoredMessage message) throws MessagingException {
+            Address[] addresses = message.getMimeMessage().getRecipients(Message.RecipientType.CC);
+            if (addresses == null || addresses.length == 0) {
+                return null;
+            }
+            return String.valueOf(addresses[0]);
+        }
+
+        private int doCompare(Comparable c1, Comparable c2) {
+            int multiplier = reverse.getAndSet(false) ? -1 : 1;
+            if (c1 == c2) {
+                return 0;
+            } else if (c1 == null) {
+                return multiplier * 1;
+            } else if (c2 == null) {
+                return multiplier * -1;
+            }
+            return multiplier * c1.compareTo(c2);
+        }
+    }
+
+
+    private class SortCommandParser extends CommandParser {
+        public SortTerm sortTerm(ImapRequestLineReader request) throws ProtocolException {
+            SortTerm sortTerm = new SortTerm();
+
+            /* Sort criteria */
+            char next = request.nextChar();
+            StringBuilder sb = new StringBuilder();
+            boolean buffering = false;
+            while (next != '\n') {
+                if (next != '(' && next != ')' && next != 32 /* space */ && next != 13 /* cr */) {
+                    sb.append(next);
+                } else {
+                    if (buffering) {
+                        sortTerm.sortCriteria.add(SortKey.valueOf(sb.toString()));
+                        sb = new StringBuilder();
+                    } else {
+                        buffering = next == '(';
+                    }
+                }
+                request.consume();
+                if (next == ')') {
+                    break;
+                }
+                next = request.nextChar();
+            }
+
+            /* Charset */
+            sortTerm.charset = consumeWord(request, new ATOM_CHARValidator());
+
+            /* Search term */
+            sortTerm.searchTerm = searchCommandParser.searchTerm(request);
+            searchCommandParser.endLine(request);
+
+            return sortTerm;
+        }
+    }
+
+    private class SortTerm {
```

Comment:
This should be split out into its own public class. This inner class is difficult to unit test.